presto-main/src/main/java/com/facebook/presto/operator/project/MergingPageOutput.java

```diff
@@ -0,0 +1,172 @@
+import com.facebook.presto.spi.type.Type;
+import com.google.common.collect.ImmutableList;
+import org.openjdk.jol.info.ClassLayout;
+
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.NotThreadSafe;
+
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Queue;
+
+import static com.facebook.presto.spi.block.PageBuilderStatus.DEFAULT_MAX_PAGE_SIZE_IN_BYTES;
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * This class is intended to be used right after the PageProcessor to ensure
+ * that the size of the pages returned by FilterAndProject and ScanFilterAndProject
+ * is big enough to do not introduce considerable synchronization overhead
+ * <p>
+ * As long as the input page contains more than {@link MergingPageOutput#minRowCount} rows
+ * or bigger than {@link MergingPageOutput#minPageSizeInBytes} it is returned as is without
+ * additional memory copy.
+ * <p>
+ * The page data that has been buffered so far before receiving a "big" page is being flushed
+ * before transferring a "big" page.
+ * <p>
+ * Although it is still possible that the {@link MergingPageOutput} may return a tiny page,
+ * this situation is considered to be rare due to the assumption that filter selectivity may not
+ * vary a lot based on the particular input page.
+ * <p>
+ * Considering the CPU time required to process(filter, project) a full (~1MB) page returned by a
+ * connector, the CPU cost of memory copying (< 50kb, < 1024 rows) is supposed to be negligible
+ */
+@NotThreadSafe
+public class MergingPageOutput
+{
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(ArrayBlockBuilder.class).instanceSize();
+
+    private final List<Type> types;
+    private final PageBuilder pageBuilder;
+    private final Queue<PageProcessorOutput> inputQueue = new LinkedList<>();
+    private final Queue<Page> outputQueue = new LinkedList<>();
+
+    private final long minPageSizeInBytes;
+    private final long minRowCount;
+
+    private boolean finishing = false;
+
+    public MergingPageOutput(Iterable<? extends Type> types, int minPageSizeInBytes, int minRowCount)
+    {
+        this(types, minPageSizeInBytes, minRowCount, DEFAULT_MAX_PAGE_SIZE_IN_BYTES);
+    }
+
+    public MergingPageOutput(Iterable<? extends Type> types, int minPageSizeInBytes, int minRowCount, int maxPageSizeInBytes)
+    {
+        this.types = ImmutableList.copyOf(requireNonNull(types, "types is null"));
+        this.minPageSizeInBytes = minPageSizeInBytes;
+        this.minRowCount = minRowCount;
+        pageBuilder = PageBuilder.withMaxPageSize(maxPageSizeInBytes, this.types);
+    }
+
+    public boolean needsInput()
+    {
+        return !finishing && outputQueue.isEmpty() && inputQueue.isEmpty();
+    }
+
+    public void addInput(PageProcessorOutput input)
+    {
+        requireNonNull(input, "input is null");
```

Comment:
Add `checkState(inputQueue.isEmpty()`.
Or rather `checkState(!processedInput.isPresent()` if you look at my comment above.